### PR TITLE
store: handle v2 error when fetching assertions

### PR DESCRIPTION
**Pull request by stolowski**
Prepare assertion fetching for v2 errors (we won't hit those unless the endpoint url is switched to v2 as well).
RFC, no unit tests yet, will add them once the change is ack'ed.

### CODE REVIEW

File: store/store_asserts.go [hunk 1]
```diff
@@ -66,6 +76,7 @@ func (s *Store) Assertion(assertType *asserts.AssertionType, primaryKey []string
 		asrt, e = dec.Decode()
 		return e
 	}, func(svcErr *assertionSvcError) error {
+		// applicable to v1 errors only, no Status in v2 error response.
```
Comment (pedronis) [inline store/store_asserts.go:79]: v1 has "error_list" not "error-list" I think, so the code could be

```
(len(svcErr.ErrorList) > 0 && svcErr.ErrorList[0].Code == "not-found" /* v2 errors */ ) || svcErr.Status == 404 /* v1 */
```


no ?

File: store/store_asserts.go [hunk 2]
```diff
@@ -109,6 +129,14 @@ func (s *Store) downloadAssertions(u *url.URL, decodeBody func(io.Reader) error,
 						return e
 					}
 				}
+				// default error handling
+
+				// is it v2 error?
+				if svcErr.Status == 0 && len(svcErr.ErrorList) > 0 {
```
Comment (pedronis) [inline store/store_asserts.go:135]: this can be just len(svcErr.ErrorList) > 0

File: store/store_asserts.go [hunk 3]
```diff
@@ -109,6 +129,14 @@ func (s *Store) downloadAssertions(u *url.URL, decodeBody func(io.Reader) error,
...
+					return fmt.Errorf("assertion service error: [%s] %q", svcErr.ErrorList[0].Code, svcErr.ErrorList[0].Message)
```
Comment (pedronis) [inline store/store_asserts.go:136]: the v2 errors contain the title in the message, so we'll get:

"[not-found] not found: ..."

we should either just use Message and not code?  or strip title from message and putting into "[...]"

**Review by pedronis — COMMENTED** [review]
did a pass, let me know if you have questions, this is slightly ugly on the originating side, not much we can do

File: store/store_asserts.go (hunk 1 above)
Comment (pedronis) [reply store/store_asserts.go:79]: an option might also be to stick some of this logic to assertionSvcError as methods

Comment (stolowski) [reply store/store_asserts.go]: Ah yes, I missed "_" vs "-", thanks, that indeed simplifies the check.

File: store/store_asserts.go [hunk 4]
```diff
@@ -76,17 +76,8 @@ func (s *Store) Assertion(assertType *asserts.AssertionType, primaryKey []string
...
-			headers, _ := asserts.HeadersFromPrimaryKey(assertType, primaryKey)
-			return &asserts.NotFoundError{
-				Type:    assertType,
-				Headers: headers,
-			}
-		}
-		// v2 error response, only error-list is present.
-		if svcErr.Status == 0 && len(svcErr.ErrorList) > 0 && svcErr.ErrorList[0].Code == "not-found" {
+		// error-list indicates v2 error response.
+		if (len(svcErr.ErrorList) > 0 && svcErr.ErrorList[0].Code == "not-found" /* v2 error */) || svcErr.Status == 404 {
```
Comment (pedronis) [inline store/store_asserts.go:80]: maybe introduce  svcErr.isNotFound() for this long predicate?

File: store/store_asserts.go [hunk 5]
```diff
@@ -132,8 +123,8 @@ func (s *Store) downloadAssertions(u *url.URL, decodeBody func(io.Reader) error,
 				// default error handling
 
 				// is it v2 error?
```
Comment (pedronis) [inline store/store_asserts.go:125]: maybe make turn this code into:  assertsionSvcError.toError() returning an error

**Review by pedronis — APPROVED** [review]
thank you,  main suggestion given the new complexity is to hide some of it into helpers on the assertionSvcError, it also mean when we drop v1 support we don't have to touch the main code again but only the helpers

**Review by mvo5 — APPROVED** [review]
LGTM

**Pull request merged**